### PR TITLE
Added docstring for `grad`

### PR DESCRIPTION
**Pull request by vfdev-5**
Description:
- Addresses #224
- Added docs for grad. 
- Cleaned unused imports and other nits

### CODE REVIEW

**Comment by vfdev-5** [comment]
@samdow sounds good !

> >  By the way we can omit typehints in docstring and use a sphinx plugin which would get the typehint from the code

> Noticed that you did this for grad and this is great!

Actually, I did put typehints to the docstring. For example

```python
func (Callable): A Python function that takes one or more arguments.
```

File: functorch/_src/eager_transforms.py [hunk 1]
```diff
@@ -324,14 +322,89 @@ def wrapper(*args, **kwargs):
...
+        In this case, ``grad(f)(x)`` will respect the inner ``torch.no_grad``.
+
+        Case 2: Using ``grad`` inside ``torch.no_grad`` context manager:
+
+            >>> with torch.no_grad():
+            >>>     grad(f)(x)
+
+        In this case, ``grad`` will respect the inner ``torch.no_grad``, but not the
+        outer one. This is because ``grad`` is a "function transform": its result
+        should not depend on the result of a context manager outside of ``f``.
```
Comment (samdow) [inline functorch/_src/eager_transforms.py:398]: Thanks for copying this note up!

File: functorch/_src/eager_transforms.py [hunk 2]
```diff
@@ -324,14 +322,89 @@ def wrapper(*args, **kwargs):
...
+    with respect to the input. This operator can be used multiple times to compute
+    higher-order gradients.
+
+    Args:
+        func (Callable): A Python function that takes one or more arguments.
+            Must return a single-element Tensor. If specified ``has_aux=True``, function can
+            return a tuple of single-element Tensor and other auxiliairy objects:
+            ``(output, aux)``.
+        argnums (int or Tuple[int]): Specifies arguments to compute gradients with respect to.
+            ``argnums`` can be single integer or tuple of integers. Default: 0.
```
Comment (samdow) [inline functorch/_src/eager_transforms.py:337]: nit:
```suggestion
            :attr:`argnums` can be single integer or tuple of integers. Default: 0.
```

File: functorch/_src/eager_transforms.py [hunk 3]
```diff
@@ -324,14 +322,89 @@ def wrapper(*args, **kwargs):
...
+        has_aux (bool): Flag indicating that ``func`` returns a tensor and other
```
Comment (samdow) [inline functorch/_src/eager_transforms.py:338]: nit: 
```suggestion
        has_aux (bool): Flag indicating that :attr:`func` returns a tensor and other
```

File: functorch/_src/eager_transforms.py [hunk 4]
```diff
@@ -324,14 +322,89 @@ def wrapper(*args, **kwargs):
...
+            auxiliairy objects: ``(output, aux)``. Default: False.
+
+    Returns:
+        Function to compute gradients on its inputs. By default, the output of
+        the function is gradients tensor. If specified ``has_aux=True``, tuple of gradients
+        and output auxiliairy objects is returned.
+        If ``argnums`` is a tuple of integers, a tuple of output gradients with respect to
```
Comment (samdow) [inline functorch/_src/eager_transforms.py:345]: nit: 
```suggestion
        If :attr:`argnums` is a tuple of integers, a tuple of output gradients with respect to
```

File: functorch/_src/eager_transforms.py [hunk 5]
```diff
@@ -324,14 +322,89 @@ def wrapper(*args, **kwargs):
...
+        each ``argnums`` value is returned
```
Comment (samdow) [inline functorch/_src/eager_transforms.py:346]: nit:
```suggestion
        each :attr:`argnums` value is returned
```

File: functorch/_src/eager_transforms.py [hunk 6]
```diff
@@ -324,14 +322,89 @@ def wrapper(*args, **kwargs):
...
+    Example of using ``grad``:
+
+        >>> from functorch import grad
+        >>> x = torch.randn([])
+        >>> cos_x = grad(lambda x: torch.sin(x))(x)
+        >>> assert torch.allclose(cos_x, x.cos())
+        >>>
+        >>> # Second-order gradients
+        >>> neg_sin_x = grad(grad(lambda x: torch.sin(x)))(x)
+        >>> assert torch.allclose(neg_sin_x, -x.sin())
```
Comment (samdow) [inline functorch/_src/eager_transforms.py:357]: Thanks for including the second order gradient example!

File: functorch/_src/eager_transforms.py [hunk 7]
```diff
@@ -324,14 +322,89 @@ def wrapper(*args, **kwargs):
...
+        >>> def compute_loss(weights, example, target):
+        >>>     y = model(weights, example)
+        >>>     return ((y - target) ** 2).mean()  # MSELoss
+        >>>
+        >>> weights = torch.randn(feature_size, requires_grad=True)
+        >>> examples = torch.randn(batch_size, feature_size)
+        >>> targets = torch.randn(batch_size)
+        >>> inputs = (weights,examples, targets)
+        >>> grad_weight_per_example = vmap(grad(compute_loss), in_dims=(None, 0, 0))(*inputs)
+
```
Comment (samdow) [inline functorch/_src/eager_transforms.py:378]: Could we add an example demonstrating the `has_aux` flag? Like `in_dims`, the example can be more contrived than the other examples if need be just so a new user can see using that flag

File: functorch/_src/eager_transforms.py [hunk 8]
```diff
@@ -324,14 +322,89 @@ def wrapper(*args, **kwargs):
...
+        >>>     return feature_vec.dot(weights).relu()
+        >>>
+        >>> def compute_loss(weights, example, target):
+        >>>     y = model(weights, example)
+        >>>     return ((y - target) ** 2).mean()  # MSELoss
+        >>>
+        >>> weights = torch.randn(feature_size, requires_grad=True)
+        >>> examples = torch.randn(batch_size, feature_size)
+        >>> targets = torch.randn(batch_size)
+        >>> inputs = (weights,examples, targets)
```
Comment (samdow) [inline functorch/_src/eager_transforms.py:376]: nit: missing space after weight
```suggestion
        >>> inputs = (weights, examples, targets)
```

File: functorch/_src/eager_transforms.py [hunk 9]
```diff
@@ -324,14 +322,89 @@ def wrapper(*args, **kwargs):
...
+        >>> # Second-order gradients
+        >>> neg_sin_x = grad(grad(lambda x: torch.sin(x)))(x)
+        >>> assert torch.allclose(neg_sin_x, -x.sin())
+
+    When composed with ``vmap``, ``grad`` can be used to compute per-sample-gradients:
+
+        >>> from functorch import vmap
+        >>> batch_size, feature_size = 3, 5
+        >>>
+        >>> def model(weights,feature_vec):
```
Comment (samdow) [inline functorch/_src/eager_transforms.py:364]: nit: space in between parameters
```suggestion
        >>> def model(weights, feature_vec):
```

File: functorch/_src/eager_transforms.py [hunk 10]
```diff
@@ -324,14 +322,89 @@ def wrapper(*args, **kwargs):
...
+        >>> from functorch import grad
+        >>> x = torch.randn([])
+        >>> cos_x = grad(lambda x: torch.sin(x))(x)
+        >>> assert torch.allclose(cos_x, x.cos())
+        >>>
+        >>> # Second-order gradients
+        >>> neg_sin_x = grad(grad(lambda x: torch.sin(x)))(x)
+        >>> assert torch.allclose(neg_sin_x, -x.sin())
+
+    When composed with ``vmap``, ``grad`` can be used to compute per-sample-gradients:
```
Comment (samdow) [inline functorch/_src/eager_transforms.py:359]: This is a really great example--it feels like something users can immediately pull down and start modifying and building off of.

File: functorch/_src/eager_transforms.py [hunk 11]
```diff
@@ -324,14 +322,89 @@ def wrapper(*args, **kwargs):
...
+        >>>
+        >>> def compute_loss(weights, example, target):
+        >>>     y = model(weights, example)
+        >>>     return ((y - target) ** 2).mean()  # MSELoss
+        >>>
+        >>> weights = torch.randn(feature_size, requires_grad=True)
+        >>> examples = torch.randn(batch_size, feature_size)
+        >>> targets = torch.randn(batch_size)
+        >>> inputs = (weights,examples, targets)
+        >>> grad_weight_per_example = vmap(grad(compute_loss), in_dims=(None, 0, 0))(*inputs)
```
Comment (samdow) [inline functorch/_src/eager_transforms.py:377]: Could we either add another example or adjust this one to demonstrate using `in_dims`? Like `has_aux`, the example can be more contrived than the other examples if there's no good one just so a new user can see how it's used

File: functorch/_src/eager_transforms.py [hunk 12]
```diff
@@ -324,14 +322,89 @@ def wrapper(*args, **kwargs):
...
+            Must return a single-element Tensor. If specified ``has_aux=True``, function can
+            return a tuple of single-element Tensor and other auxiliairy objects:
+            ``(output, aux)``.
+        argnums (int or Tuple[int]): Specifies arguments to compute gradients with respect to.
+            ``argnums`` can be single integer or tuple of integers. Default: 0.
+        has_aux (bool): Flag indicating that ``func`` returns a tensor and other
+            auxiliairy objects: ``(output, aux)``. Default: False.
+
+    Returns:
+        Function to compute gradients on its inputs. By default, the output of
```
Comment (samdow) [inline functorch/_src/eager_transforms.py:342]: nit:
```suggestion
        Function to compute gradients with respect to its inputs. By default, the output of
```

Also, although we've mentioned it before, it would be good to note that the returned function takes the same inputs as the original in case a user misses it once

File: functorch/_src/eager_transforms.py [hunk 13]
```diff
@@ -324,14 +322,89 @@ def wrapper(*args, **kwargs):
...
+        the function is gradients tensor. If specified ``has_aux=True``, tuple of gradients
```
Comment (samdow) [inline functorch/_src/eager_transforms.py:343]: Maybe
```suggestion
        the function is the gradient tensor with respect to the first argument If specified ``has_aux=True``, tuple of gradients
```

**Review by samdow — APPROVED** [review]
This looks really great. I had some wording suggestions (glad to discuss if you disagree with any) + nits (it seems from the PyTorch docs that `` :attr:`<attr name>` `` is preferred over ` ``<attr name>`` ` ). I also requested a couple more examples, but if you don't want to add those, we can add them in a separate PR

**Comment by vfdev-5** [comment]
Thanks for the review @samdow , I agree with your suggestions! Examples I added here are adaptations from readme and comments in the code. I'll add few others as suggested.

File: functorch/_src/eager_transforms.py [hunk 14]
```diff
@@ -324,14 +322,89 @@ def wrapper(*args, **kwargs):
         return grad_input, output
     return wrapper
 
-def grad(f, argnums=0, has_aux=False):
-    @wraps(f)
+def grad(func: Callable, argnums: argnums_t = 0, has_aux: bool = False) -> Callable:
+    """``grad`` operator helps computing gradients of the output of a function.
```
[the remark this answers is not in the document]
Comment (vfdev-5) [reply functorch/_src/eager_transforms.py:326]: I agree, maybe we can merge this and below sentences in a single better one.

[the remark this answers is not in the document]
Comment (samdow) [reply functorch/_src/eager_transforms.py:326]: Agreed! Let me know if you want help wordsmithing it

File: functorch/_src/eager_transforms.py (hunk 7 above)
Comment (vfdev-5) [reply functorch/_src/eager_transforms.py:377]: Added an example of using `has_aux` and `argnums` together. @samdow what do you think?

Comment (samdow) [reply functorch/_src/eager_transforms.py:377]: Looks awesome! Thanks for adding it!

**Review by samdow — APPROVED** [review]
Thanks for all the updates! Looks great

**Comment by vfdev-5** [comment]
@samdow can we merge this PR or do you want me to add something else here ?

**Comment by samdow** [comment]
Sorry I didn't realize that I had to merge it. Thanks again for adding this!

**Comment by vfdev-5** [comment]
I do not have (yet) write rights for this repository :) that's why I asked you.
